bittide/src/Clash/Hitl.hs

```diff
@@ -0,0 +1,289 @@
+--
+testsFromEnum :: (Show a, Bounded a, Enum a) => (a -> Probes a) -> HitlTests a
+testsFromEnum f = Map.fromList $ map (\a -> (Text.pack (show a), f a)) [minBound..]
+
+-- | A list, but with a custom "ToJSON" instance to work around Vivado issues
+newtype PackedList a = PackedList [a]
+
+-- | XXX: Custom "ToJSON" instance for "PackedList" that converts an empty
+--        "PackedList" into a 'Aeson.Null' to accomodate Vivado's poorly
+--        implemented JSON/YAML parser.
+instance ToJSON a => ToJSON (PackedList a) where
+  toJSON (PackedList []) = Aeson.Null
+  toJSON (PackedList l) = toJSON l
+
+-- | A map from a probe name to a (binary) value with a custom "ToJSON" instance
+-- to work around Vivado issues.
+data PackedProbes = PackedProbes (Map Text Natural)
+
+-- | XXX: Custom "ToJSON" instance for "PackedProbes" that converts an empty
+--        "PackedProbes" into a 'Aeson.Null' to accomodate Vivado's poorly
+--        implemented JSON/YAML parser.
+instance ToJSON PackedProbes where
+  toJSON (PackedProbes []) = Aeson.Null
+  toJSON (PackedProbes l) = toJSON l
+
+-- | See "PackedTests"
+data PackedTargetRef
+  = ByIndex { index :: Integer }
+  deriving (Generic, ToJSON)
+
+-- | See "PackedTests"
+data PackedTarget = PackedTarget
+  { target :: PackedTargetRef
+  , probes :: PackedProbes
+  }
+  deriving (Generic, ToJSON)
+
+-- | See "PackedTests"
+data PackedTest = PackedTest
+  { targets :: PackedList PackedTarget
+  }
+  deriving (Generic, ToJSON)
+
+-- | Intermediate representation of "HitlTests". There to provide trivial instances
+-- of "ToJSON".
+data PackedTests = PackedTests
+  { defaults :: PackedProbes
+  , tests :: Map Text PackedTest
+  }
+
+instance ToJSON PackedTests where
+  toJSON (PackedTests{defaults, tests}) = object
+    [ ("defaults" .= object ["probes" .= defaults])
+    , "tests"     .= toJSON tests
+    ]
+
+-- | Convert an \"unpacked\" "HitlTests" to a packed version. The packed version
+-- is convertable to JSON, which in turn can be interpreted by the @HardwareTest.tcl@.
+toPacked :: forall a. BitPack a => HitlTests a -> PackedTests
+toPacked tests =
+  -- If @a@ is a zero-width type, we don't want to generate any data probes
+  case natToInteger @(BitSize a) of
+    0 -> PackedTests
+      { defaults=PackedProbes []
+      , tests=fromList [(name, go0 probes) | (name, probes) <- toList tests]
+      }
+    _ -> PackedTests
+      { defaults=PackedProbes [("probe_test_data", 0)]
+      , tests=fromList [(name, goN probes) | (name, probes) <- toList tests]
+      }
+ where
+  go0 probes = PackedTest
+    { targets=PackedList
+      [PackedTarget{target=ByIndex (toInteger id_), probes=PackedProbes []} | (id_, _) <- probes]
+    }
+
+  goN probes = PackedTest
+    { targets=PackedList
+      [ PackedTarget
+          { target=ByIndex (toInteger id_)
+          , probes=PackedProbes [("probe_test_data", BitVector.unsafeToNatural (pack dat))]
+          }
+      | (id_, dat) <- probes]
+    }
```

Comment:
Why not:
```hs
-- | Convert an \"unpacked\" "HitlTests" to a packed version. The packed version
-- is convertable to JSON, which in turn can be interpreted by the @HardwareTest.tcl@.
toPacked :: forall a. BitPack a => HitlTests a -> PackedTests
toPacked hitlTests = PackedTests {defaults, tests}
 where
  bs = natToInteger @(BitSize a)
  defaults = PackedProbes (if bs == 0 then [] else [("probe_test_data", 0)])
  tests=fromList [(name, go probes) | (name, probes) <- toList hitlTests]

  go probes = PackedTest {targets=PackedList targetList}
   where
    targetList
      | bs == 0 =
        [PackedTarget
          { target=ByIndex (toInteger id_)
          , probes=PackedProbes []
          } | (id_, _) <- probes]

      | otherwise =
        [PackedTarget
          { target=ByIndex (toInteger id_)
          , probes=PackedProbes [("probe_test_data", BitVector.unsafeToNatural (pack dat))]
          } | (id_, dat) <- probes
        ]
    }

Reply:
I don't like nested `where`s. See currently style, which is a compromise. Ack?